File: prolog1_examples/tree.c
#include <stdio.h>

enum treetype {
    Node,
    Leaf
};

struct tree {
    enum treetype type;
    union {
        struct {
            struct tree *left;
            struct tree *right;
        } node;
        struct {
            int value;
        } leaf;
    } u;
};


int sum_tree(struct tree *tree)
{
    switch(tree->type) {
    case Leaf:
        return tree->u.leaf.value;
    case Node:
        return sum_tree(tree->u.node.left) + sum_tree(tree->u.node.right);
    }
}

int sum_tree2(struct tree *tree)
{
  int sum = 0; 

 repeat:
  switch(tree->type) {
  case Leaf:
    sum += tree->u.leaf.value;
    return sum;
  case Node:
    sum += sum_tree(tree->u.node.left);
    tree = tree->u.node.right;
    goto repeat;
  }
}


int main()
{
    struct tree n1, n2;
    struct tree l1, l2, l3;
    int sum;

    l1.type = l2.type = l3.type = Leaf;
    l1.u.leaf.value = 5;
    l2.u.leaf.value = 3;
    l3.u.leaf.value = 2;

    n1.type = n2.type = Node;
    n1.u.node.left = &l1;
    n1.u.node.right = &n2;
    n2.u.node.left = &l2;
    n2.u.node.right = &l3;
    
    sum = sum_tree(&n1);
    printf("sum is: %i\n", sum);
    
    return 0;
}

File: prolog1_examples/tree.pl
% :- type tree ---> leaf(int) | node(tree, tree).

% tree_sum(+Tree, ?Sum): A Tree fa leveleiben levõ
% értékek összege Sum.
tree_sum(leaf(Value), Value).
tree_sum(node(Left, Right), S) :-
	tree_sum(Left, S1),
	tree_sum(Right, S2),
	S is S1+S2.























% tree_sum2(?Tree, ?Sum): A Tree fa leveleiben levõ
% értékek összege (a plus/3 eljárás szerint) Sum.
tree_sum2(leaf(Value), Value).
tree_sum2(node(Left, Right), S) :-
	plus(SL, SR, S),
	tree_sum2(Left, SL),
	tree_sum2(Right, SR).


% plus(X, Y, Z): X + Y = Z, X > 0, Y > 0, Z =< 5.
plus(1,1,2).
plus(1,2,3).
plus(1,3,4).
plus(1,4,5).
plus(2,1,3).
plus(2,2,4).
plus(2,3,5).
plus(3,1,4).
plus(3,2,5).
plus(4,1,5).


test :-
	tree_sum(node(leaf(1),
		      node(leaf(2),
			   leaf(2))), S),
	write(S), nl, fail.
test.

test2(S) :-
	tree_sum2(T, S),
	write(T), nl, fail.
test2(_).


